### PR TITLE
vectorize decorator

### PR DESCRIPTION
**Pull request by lintusj1**
```
@elfi.tools.vectorize
def simulator(a, b, random_state=None):
    # Simulator code
    pass
```

### CODE REVIEW

File: elfi/tools.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+def run_vectorized(operation, *inputs, constants=None, batch_size=None, **kwargs):
+    """Runs the operation as if it was vectorized over the individual runs in the batch.
+
+    Helper for cases when you have an operation that does not support vector arguments.
+
+    Parameters
+    ----------
+    operation : callable
+        operation that will be run `batch_size` times.
+    inputs
```
Comment (vuolleko) [inline elfi/tools.py]: Missing explanation

File: elfi/tools.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    constants = [] if constants is None else list(constants)
+
+    # Check input and set constants and batch_size if needed
+    for i, inpt in enumerate(inputs):
+        if i in constants:
+            continue
+
+        try:
+            l = len(inpt)
+        except:
```
Comment (vuolleko) [inline elfi/tools.py]: This could fail also because of user error.

File: elfi/tools.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            constants.append(i)
+            l = 1
+
+        if l != 1:
+            if batch_size is None:
+                batch_size = l
+            elif batch_size != l:
```
Comment (vuolleko) [inline elfi/tools.py]: Why is the user even allowed to give batch_size?

Comment (lintusj1) [inline elfi/tools.py]: The batch size would come from the framework. But not all nodes receive it as input, so in that case we try to infer it.

**Pull request merged**